# Profitable App Profiles for the App Store and Google Play Markets.


Our company builds Android and iOS mobile apps and We make our apps available on Google Play and the App Store for the users . Moreover We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means our revenue for any given app is mostly influenced by the number of users who use our app — the more users that see and engage with the ads, the better. 

Our aim in this project is to find mobile app profiles that are profitable for the App Store and Google Play markets by analysing the data to help our developers understand what kind of apps are likely to attract more users.

## Opening and Exploring the Datasets.


 * The first dataset contains data about approximately 10,000 Android apps from Google Play. Here is the link to download the dataset [GooglePlaystore](https://www.kaggle.com/lava18/google-play-store-apps)
 * The second dataset contains data about  approximately 7,000 iOS apps from the App Store. Here is the link to download the dataset [AppStore](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

Let's start by opening and exploring these two data sets.

In [49]:
from csv import reader

# Reading Google PLay Store Data 
opened_file = open('googleplaystore.csv', encoding = "utf8" )
read_file = reader(opened_file)
playstore = list(read_file)
playstore_header = playstore[0]
playstore_data = playstore[1:]

# Reading App Store Data 
opened_file = open('AppleStore.csv', encoding = "utf8")
read_file = reader(opened_file)
appstore = list(read_file)
appstore_header = appstore[0]
appstore_data = appstore[1:]

To make it a bit easier to explore the data sets, we'll first write a function named **explore_data()**  

In [50]:
def explore_dataset(dataset, start, end, rows_and_columns=True):
    data = dataset[start:end]
    for row in data:
        print(row)
        print('\n')
    if rows_and_columns:
        print(f'Number of Rows : {len(dataset)}') 
        print(f'Number of Columns: {len(dataset[0])}')

# dataset - list of lists
# start - starting indices for slicing the dataset
# end  - ending indices for slicing the dataset
# rows_and_column = bool, default (True)

Let's first explore the Google Play Store Data

In [51]:
print(playstore_header,'\n')

explore_dataset(playstore_data, 1, 5, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of Rows : 10841
Number of Columns: 13


we can see that the Google Play Store data set contains data of 10841 apps across 13 columns (wherein each column is a feature of these apps).

At first glance, few features that seem interesting are- 'App', 'Category','Rrating', 'Reviews', 'Installs', 'Type', 'Price', and 'Genres'.

Now let's explore the App Store data set.

In [52]:
print(appstore_header, '\n')

explore_dataset(appstore_data, 1, 5, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of Rows : 7197
Number of Columns: 16


we can see that the App Store data set contains data of 7197 apps across 16 columns (wherein each column is a feature of these apps).

At first glance few, features that seem interesting are - 'track_name', 'currency', 'price', 'rating_count_tot', 'user_rating', and 'prime_genre'.

Please refer to this [Documentation](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home) to get details regarding each column. 

Before beginning our analysis, we need to make sure the data we analyze is accurate, otherwise the results of our analysis will be wrong. This means that we need to:

Detect inaccurate data, and correct or remove it.
Detect duplicate data, and remove the duplicates.

# Data Cleaning
The Google Play data set has a dedicated [discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion) section, and we can see that [one of the discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) describes an error for a certain row ( row 10472).



In [53]:
print(playstore_header, '\n')
print(playstore_data[0], '\n')
print(playstore_data[10472])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


By cross-referencing the data with acorrect data and the header we can find that for the app *'Life Made WI-Fi Touchscreen Photo Frame'* the rating is 19 which is clearly off because the maximum rating allowed in the Google Play Store is 5 . 
This problem is caused due to missing data in the 'category' column , hence we can delete this row .

In [54]:
print(f'Total rows before deleting incorrect data: {len(playstore_data)}', '\n')

# deleting incorrect data
del(playstore_data[10472])

print(f'Total rows after deleting incorrect data: {len(playstore_data)}')

Total rows before deleting incorrect data: 10841 

Total rows after deleting incorrect data: 10840


By going through the dedicated [discussion](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion) section of the App Store data set we dont find any incorrect data being reported hence we need not remove any data from this data set.

## Removing duplicate data.

The Google Play data set has a dedicated [discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion) section, and we can see that there are instances of duplicate data being reported.

### Part 1

In [55]:
def find_duplicates(dataset, no_of_dupicates):
    duplicate_apps = []
    unique_apps = []
    for row in dataset:
        app_name = row[0]
        if app_name in unique_apps:
            duplicate_apps.append(app_name)
        else:
            unique_apps.append(app_name)

    print(f'Number of duplicate apps: {len(duplicate_apps)}')
    print('')
    print(f'Example of duplicate apps :\n {[duplicate_apps[:no_of_dupicates]]}')

In [56]:
find_duplicates(playstore_data, 20)

Number of duplicate apps: 1181

Example of duplicate apps :
 [['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express']]


From above output it is clear that in the Google Play Store data set in total ,there are 1181 cases wherein the apps occur more than once.

We don't want to count certain apps more than once when we analyze data, so we need to remove the duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, but we could probably find a better way.

Let's futher see why these apps are occuring more than once

In [57]:
# example 
for row in playstore_data:
    app_name = row[0]
    if app_name == 'Slack':
        print(row)

['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


By printing an example of a duplicate app (Slack) we can see that the Slack app has been recorded 3 times 
The differnce between the first record and the third record can be seen in the 4th column that is the 'reviews'.
The difference in numbers denote that the data was collected at different times. We can use this as a criteria for keeping the appropriate record.
We won't remove rows randomly, but rather we'll keep the rows that have the highest number of reviews because the higher the number of reviews, the more recent and reliable the ratings are. 

To do that, we will:

* Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
* Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)

In [58]:
max_reviews = {}

for row in playstore_data:
    app_name = row[0]
    no_reviews = float(row[3])
    if (app_name in max_reviews) and (max_reviews[app_name] < no_reviews):
        max_reviews[app_name] = no_reviews
    elif app_name not in max_reviews:
        max_reviews[app_name] = no_reviews

In a previous code cell, we found that there are 1,181 cases where an app occurs more than once, so the length of our dictionary (of unique apps) should be equal to the difference between the length of our data set and 1,181.

In [59]:
print(f'Expected Length: {len(playstore_data) - 1181}')
print(f'Actual Length: {len(max_reviews)}')

Expected Length: 9659
Actual Length: 9659


Now, let's use the reviews_max dictionary to remove the duplicates. For the duplicate cases, we'll only keep the entries with the highest number of reviews. In the code cell below:

* We start by initializing two empty lists, android_clean and             already_added.
* We loop through the android data set, and for every iteration:

   - We isolate the name of the app and the number of reviews.
   - We add the current row (app) to the android_clean list, and the         app name (name) to the already_added list if:
       - The number of reviews of the current app matches the number              of reviews of that app as described in the reviews_max                  dictionary; and
       - The name of the app is not already in the already_added                  list. We need to add this supplementary condition to                    account for those cases where the highest number of reviews              of a duplicate app is the same for more than one entry (for              example, the Box app has three entries, and the number of                reviews is the same). If we just check for                              reviews_max[name] == n_reviews, we'll still end up with                  duplicate entries for some apps.

In [60]:
playstore_no_duplicates = []
already_added = []

for row in playstore_data:
    app_name = row[0]
    no_reviews = float(row[3])
    if (max_reviews[app_name] == no_reviews) and (app_name not in already_added):
        playstore_no_duplicates.append(row)
        already_added.append(app_name)

In [61]:
explore_dataset(playstore_no_duplicates, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of Rows : 9659
Number of Columns: 13


## Removing Non-English Apps 

English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;, etc.), and other symbols (+, *, /, etc.). All these characters that are specific to English texts are encoded using the ASCII standard. Each ASCII character has a corresponding number between 0 and 127 associated with it, and we can take advantage of that to build a function that checks an app name and tells us whether it contains non-ASCII characters.

In [62]:
def english_app_check(string):
    for character in string:
        if ord(character) > 127:
            return False
        else:
            return True

In [63]:
print(english_app_check('Instagram'))
print(english_app_check('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english_app_check('Docs To Go™ Free Office Suite'))
print(english_app_check('Instachat 😜'))

True
False
True
True


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we’ll remove useful apps if we use the function in its current form. To minimize the impact of data loss, we’ll only remove an app if its name has more than three non-ASCII characters:

In [64]:
def english_app_check(string):
    non_ascii = 0
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
    if non_ascii > 3:
        return False
    else:
        return True

In [65]:
print(english_app_check('Docs To Go™ Free Office Suite'))
print(english_app_check('Instachat 😜'))
print(english_app_check('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


The function is still not perfect, and very few non-English apps might get past our filter, but this seems good enough at this point in our analysis — we shouldn’t spend too much time on optimization at this point. Below, we use the english_app_check() function to filter out the non-English apps for both data sets:

In [66]:
# playstore

playstore_english = []

for app in playstore_no_duplicates:
    app_name = app[0]
    if english_app_check(app_name):
        playstore_english.append(app)

In [67]:
explore_dataset(playstore_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of Rows : 9614
Number of Columns: 13


In [68]:
# appstore

appstore_english = []

for app in appstore_data:
    app_name = app[1]
    if english_app_check(app_name):
        appstore_english.append(app)

In [69]:
explore_dataset(appstore_english, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of Rows : 6183
Number of Columns: 16


## Isolating the Free Apps

As we mentioned in the introduction, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps, and we’ll need to isolate only the free apps for our analysis. Below, we isolate the free apps for both our data sets.

In [70]:
# playstore

playstore_free = []
playstore_paid = []

for app in playstore_english:
    price = app[7]
    if price == '0':
        playstore_free.append(app)
    else:
        playstore_paid.append(app)

In [71]:
print(playstore_header, '\n')
explore_dataset(playstore_free, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of Rows : 8864
Number of Columns: 13


In [72]:
# appstore

appstore_free = []
appstore_paid = []

for app in appstore_english:
    price = app[4]
    if price == '0.0':
        appstore_free.append(app)
    else:
        appstore_paid.append(app)

In [73]:
print(appstore_header, '\n')
explore_dataset(appstore_free, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of Rows : 3222
Number of Columns: 16


## Most Common Apps by Genre

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

In [74]:
# Frequency Table with percentage

def freq_table(dataset, column_index):

    table = {}
    total = 0

    for row in dataset:
        total += 1
        value = row[column_index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1

    # return table, total

    table_percentage = {}

    for key in table:
        percentage = (table[key] / total) * 100
        table_percentage[key] = percentage

    return table_percentage

In [75]:
print(freq_table(appstore_free, 11))

{'Social Networking': 3.2898820608317814, 'Photo & Video': 4.9658597144630665, 'Games': 58.16263190564867, 'Music': 2.0484171322160147, 'Reference': 0.5586592178770949, 'Health & Fitness': 2.0173805090006205, 'Weather': 0.8690254500310366, 'Utilities': 2.5139664804469275, 'Travel': 1.2414649286157666, 'Shopping': 2.60707635009311, 'News': 1.3345747982619491, 'Navigation': 0.186219739292365, 'Lifestyle': 1.5828677839851024, 'Entertainment': 7.883302296710118, 'Food & Drink': 0.8069522036002483, 'Sports': 2.1415270018621975, 'Book': 0.4345127250155183, 'Finance': 1.1173184357541899, 'Education': 3.662321539416512, 'Productivity': 1.7380509000620732, 'Business': 0.5276225946617008, 'Catalogs': 0.12414649286157665, 'Medical': 0.186219739292365}


In [76]:
# displaying percentages in descending order

def display_table(dataset, index):
    table = freq_table(dataset, index)

    table_display = []

    for key in table:
        key_value_tuple = (table[key], key)
        table_display.append(key_value_tuple)

    # return table_display

    table_sorted = sorted(table_display, reverse = True)

    # return table_sorted
    
    for entry in table_sorted:
        print(entry[1], ' : ', entry[0])

We start by examining the frequency table for the prime_genre column of the App Store data set

In [77]:
display_table(appstore_free, -5)

Games  :  58.16263190564867
Entertainment  :  7.883302296710118
Photo & Video  :  4.9658597144630665
Education  :  3.662321539416512
Social Networking  :  3.2898820608317814
Shopping  :  2.60707635009311
Utilities  :  2.5139664804469275
Sports  :  2.1415270018621975
Music  :  2.0484171322160147
Health & Fitness  :  2.0173805090006205
Productivity  :  1.7380509000620732
Lifestyle  :  1.5828677839851024
News  :  1.3345747982619491
Travel  :  1.2414649286157666
Finance  :  1.1173184357541899
Weather  :  0.8690254500310366
Food & Drink  :  0.8069522036002483
Reference  :  0.5586592178770949
Business  :  0.5276225946617008
Book  :  0.4345127250155183
Navigation  :  0.186219739292365
Medical  :  0.186219739292365
Catalogs  :  0.12414649286157665


We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set. The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. 

Let’s continue by examining the Genres and Category columns of the Google Play data set

In [78]:
display_table(playstore, 1)

FAMILY  :  18.18852610219517
GAME  :  10.551558752997602
TOOLS  :  7.775318206972884
MEDICAL  :  4.270429809998156
BUSINESS  :  4.242759638443092
PRODUCTIVITY  :  3.910717579782328
PERSONALIZATION  :  3.6155690831949827
COMMUNICATION  :  3.5694521306032096
SPORTS  :  3.541781959048146
LIFESTYLE  :  3.523335178011437
FINANCE  :  3.3757609297177646
HEALTH_AND_FITNESS  :  3.1451761667589007
PHOTOGRAPHY  :  3.089835823648773
SOCIAL  :  2.7209002029145912
NEWS_AND_MAGAZINES  :  2.6102195166943365
SHOPPING  :  2.3980815347721824
TRAVEL_AND_LOCAL  :  2.3796347537354734
DATING  :  2.158273381294964
BOOKS_AND_REFERENCE  :  2.1306032097399004
VIDEO_PLAYERS  :  1.614093340712046
EDUCATION  :  1.4388489208633095
ENTERTAINMENT  :  1.3742851872348276
MAPS_AND_NAVIGATION  :  1.263604501014573
FOOD_AND_DRINK  :  1.1713705958310274
HOUSE_AND_HOME  :  0.8116583656152001
LIBRARIES_AND_DEMO  :  0.7839881940601365
AUTO_AND_VEHICLES  :  0.7839881940601365
WEATHER  :  0.7563180225050729
ART_AND_DESIGN  :  0.

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.).

## Most Popular Apps By Genre On The Appstore

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we’ll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

In [79]:
# Generating a frequency table to get the unique app genres 
appstore_genres = freq_table(appstore_free, -5)
# print(appstore_genres)

for genre in appstore_genres:
    total = 0
    len_genre = 0


    for app in appstore_free:
        app_genre = app[-5]
        if app_genre == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1

    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)



Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


In [80]:
On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together

SyntaxError: invalid syntax (<ipython-input-80-b98da675c429>, line 1)

In [34]:
for app in appstore_free:
    if app[-5] == 'Navigation':
        print(app[1], ' : ', app[5])

Waze - GPS Navigation, Maps & Real-time Traffic  :  345046
Google Maps - Navigation & Transit  :  154911
Geocaching®  :  12811
CoPilot GPS – Car Navigation & Offline Maps  :  3582
ImmobilienScout24: Real Estate Search in Germany  :  187
Railway Route Search  :  5


The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number. Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we’ll leave this level of detail for later. Reference apps have 74,942 user ratings on average, but it’s actually the Bible and Dictionary.com which skew up the average rating:

In [35]:
for app in appstore_free:
    if app[-5] == 'Social Networking':
        print(app[1], ' : ', app[5])

Facebook  :  2974676
Pinterest  :  1061624
Skype for iPhone  :  373519
Messenger  :  351466
Tumblr  :  334293
WhatsApp Messenger  :  287589
Kik  :  260965
ooVoo – Free Video Call, Text and Voice  :  177501
TextNow - Unlimited Text + Calls  :  164963
Viber Messenger – Text & Call  :  164249
Followers - Social Analytics For Instagram  :  112778
MeetMe - Chat and Meet New People  :  97072
We Heart It - Fashion, wallpapers, quotes, tattoos  :  90414
InsTrack for Instagram - Analytics Plus More  :  85535
Tango - Free Video Call, Voice and Chat  :  75412
LinkedIn  :  71856
Match™ - #1 Dating App.  :  60659
Skype for iPad  :  60163
POF - Best Dating App for Conversations  :  52642
Timehop  :  49510
Find My Family, Friends & iPhone - Life360 Locator  :  43877
Whisper - Share, Express, Meet  :  39819
Hangouts  :  36404
LINE PLAY - Your Avatar World  :  34677
WeChat  :  34584
Badoo - Meet New People, Chat, Socialize.  :  34428
Followers + for Instagram - Follower Analytics  :  28633
GroupMe  :  

In [36]:
for app in appstore_free:
    if app[-5] == 'Music':
        print(app[1], ' : ', app[5])

Pandora - Music & Radio  :  1126879
Spotify Music  :  878563
Shazam - Discover music, artists, videos & lyrics  :  402925
iHeartRadio – Free Music & Radio Stations  :  293228
SoundCloud - Music & Audio  :  135744
Magic Piano by Smule  :  131695
Smule Sing!  :  119316
TuneIn Radio - MLB NBA Audiobooks Podcasts Music  :  110420
Amazon Music  :  106235
SoundHound Song Search & Music Player  :  82602
Sonos Controller  :  48905
Bandsintown Concerts  :  30845
Karaoke - Sing Karaoke, Unlimited Songs!  :  28606
My Mixtapez Music  :  26286
Sing Karaoke Songs Unlimited with StarMaker  :  26227
Ringtones for iPhone & Ringtone Maker  :  25403
Musi - Unlimited Music For YouTube  :  25193
AutoRap by Smule  :  18202
Spinrilla - Mixtapes For Free  :  15053
Napster - Top Music & Radio  :  14268
edjing Mix:DJ turntable to remix and scratch music  :  13580
Free Music - MP3 Streamer & Playlist Manager Pro  :  13443
Free Piano app by Yokee  :  13016
Google Play Music  :  10118
Certified Mixtapes - Hip Hop 

In [38]:
for app in appstore_free:
    if app[-5] == 'Reference':
        print(app[1], ' : ', app[5])

Bible  :  985920
Dictionary.com Dictionary & Thesaurus  :  200047
Dictionary.com Dictionary & Thesaurus for iPad  :  54175
Google Translate  :  26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran  :  18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition  :  17588
Merriam-Webster Dictionary  :  16849
Night Sky  :  12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE)  :  8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools  :  4693
GUNS MODS for Minecraft PC Edition - Mods Tools  :  1497
Guides for Pokémon GO - Pokemon GO News and Cheats  :  826
WWDC  :  762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free  :  718
VPN Express  :  14
Real Bike Traffic Rider Virtual Reality Glasses  :  8
教えて!goo  :  0
Jishokun-Japanese English Dictionary & Translator  :  0


## Most Popular Apps By Genre On Playstore


For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don’t seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [41]:
display_table(playstore_free, 5)

1,000,000+  :  15.726534296028879
100,000+  :  11.552346570397113
10,000,000+  :  10.548285198555957
10,000+  :  10.198555956678701
1,000+  :  8.393501805054152
100+  :  6.915613718411552
5,000,000+  :  6.825361010830325
500,000+  :  5.561823104693141
50,000+  :  4.7721119133574
5,000+  :  4.512635379061372
10+  :  3.5424187725631766
500+  :  3.2490974729241873
50,000,000+  :  2.3014440433213
100,000,000+  :  2.1322202166064983
50+  :  1.917870036101083
5+  :  0.78971119133574
1+  :  0.5076714801444043
500,000,000+  :  0.2707581227436823
1,000,000,000+  :  0.22563176895306858
0+  :  0.04512635379061372
0  :  0.01128158844765343


One problem with this data is that is not precise. For instance, we don’t know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don’t need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don’t need perfect precision with respect to the number of users. We’re going to leave the numbers as they are, which means that we’ll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on. To perform computations, however, we’ll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We’ll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [42]:
# Generating a frequency table to get the unique app genres
playstore_categories = freq_table(playstore_free, 1)

# looping over unique app genres
for category in playstore_categories:
    total = 0
    len_category = 0

    for app in playstore_free:
        app_category = app[1]
        if app_category == category:
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1

    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)


ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:

In [44]:
for app in playstore_free:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we removed all the communication apps that have over 100 million installs, the average would be reduced by more than ten times

In [45]:
under_100_m = []

for app in playstore_free:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if float(n_installs) < 100000000:
        under_100_m.append(float(n_installs))

sum(under_100_m) / len(under_100_m)

2821347.6231027693

We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, and MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.). Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against. The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we’d like to come up with a different app recommendation if possible. The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It’s interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play. Let’s take a look at some of the apps from this genre and their number of installs

In [46]:
for app in playstore_free:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there’s still a small number of extremely popular apps that skew the average

In [47]:
for app in playstore_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. Let’s try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [48]:
for app in playstore_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

   This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it’s probably not a good idea to build similar apps since there’ll be some significant competition. We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc. 